test/web/test_server.py

```diff
@@ -43,6 +40,28 @@ def test_broken_pipe(
     assert web_session.get('/').text == ''
 
 
+@pytest.mark.xfail(reason="https://github.com/onefinestay/nameko/issues/368")
+def test_client_disconnect_os_error(
+    container_factory, web_config, web_config_port, web_session
+):
+    """ Regression for https://github.com/onefinestay/nameko/issues/368
+    """
+    container = container_factory(ExampleService, web_config)
+    container.start()
+
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s.connect(('127.0.0.1', web_config_port))
+
+    with patch.object(HttpOnlyProtocol, 'handle_one_request') as handle:
+        handle.side_effect = OSError('raw readinto() returned invalid length')
```

Comment:
It's difficult to demonstrate that something *doesn't* crash when the crash cause is non-deterministic.

I have something that seems to work though. Do you think the following is preferable?

``` python
import eventlet
import pytest
import requests
from eventlet.wsgi import Server

from nameko.testing.waiting import wait_for_call
from nameko.web.handlers import http


@pytest.fixture
def container(container_factory):

    class Service(object):
        name = "service"

        @http('GET', '/answer')
        def answer(self, request):
            return "42"

    container = container_factory(Service, {})
    container.start()

    return container


def test_foo(container):

    run = [True]

    def wait_for_os_error(args, kwargs, res, exc_info):
        if exc_info and exc_info[0] == OSError:
            del run[:]
            return True

    with eventlet.Timeout(1):
        with wait_for_call(Server, 'process_request', callback=wait_for_os_error):
            while run:
                gt = eventlet.spawn(requests.get, "http://localhost:8000/answer")
                eventlet.sleep(0.01)
                gt.kill()

    # still works
    assert requests.get("http://localhost:8000/answer").text == "42"

```